primap2/pm2io/_data_reading.py

```diff
@@ -1289,26 +1340,97 @@ def harmonize_units(
+            basic_entity = basic_entity[0]
+        else:
+            gwp_to_use = None
+            basic_entity = entity
+        # print(f"basic_entity: {basic_entity}")
         # get all units for this entity
         data_this_entity = data.loc[data[entity_col] == entity]
         units_this_entity = data_this_entity[unit_col].unique()
-        # print(units_this_entity)
-        if len(units_this_entity) > 1:
-            # need unit conversion. convert to first unit (base units have second as
-            # time that is impractical)
-            unit_to = units_this_entity[0]
-            # print("unit_to: " + unit_to)
-            for unit in units_this_entity[1:]:
-                # print(unit)
-                unit_pint = ureg[unit]
-                unit_pint = unit_pint.to(unit_to)
-                # print(unit_pint)
-                factor = unit_pint.magnitude
-                # print(factor)
-                mask = (data[entity_col] == entity) & (data[unit_col] == unit)
-                data.loc[mask, data_cols] *= factor
-                data.loc[mask, unit_col] = unit_to
+
+        if len(units_this_entity) > 1 or gwp_to_use:
+            # need unit conversion. if possible convert to Gg of substance / year
+            # print(f"Entity: {entity}, units: {units_this_entity}, GWP: {gwp_to_use}")
+
+            unit_fallback = units_this_entity[0]
+            unit_to = None  # to check
+
+            # check if conversion to native unit is possible
+            try:
+                native_unit = "Gg " + basic_entity + " / yr"
+                # print(f"Testing conversion from {ureg[unit_fallback].units} to "
+                #       f"{ureg[native_unit].units} for {basic_entity}.")
+                if gwp_to_use:
+                    if ureg(unit_fallback).is_compatible_with(
+                        ureg[native_unit], gwp_to_use
+                    ):
+                        unit_to = native_unit
+                        entity_conv = True
+                    else:
+                        unit_to = unit_fallback
+                        entity_conv = False
+                else:
+                    if ureg(unit_fallback).is_compatible_with(ureg[native_unit]):
+                        unit_to = native_unit
+                        entity_conv = False
+                    else:
+                        unit_to = unit_fallback
+                        entity_conv = False
+            except pint.UndefinedUnitError:
+                # we have a gas basket or something unknown, so no conversion to native
+                # unit
+                # print(f"Exception occured for entity {entity}")
+                unit_to = unit_fallback
+                entity_conv = False
+                native_unit = "None"
+
+            # print(f"unit_to is {unit_to}, unit_fallback is {unit_fallback}, and
+            # native_unit is {native_unit}")
+            # print(f"Using {unit_to} as target unit and entity_conv = {entity_conv}")
```

Comment:
I have moved this to a separate function to keep things tidy and easily tested.

Reply:
Good. Splitting functions is something I usually forget.